src/sinks/util/encoding/mod.rs

```diff
@@ -0,0 +1,278 @@
+//! Encoding related code.
+//!
+//! You'll find two stuctures for configuration:
+//!   * `EncodingConfig<E>`: For sinks without a default `Encoding`.
+//!   * `EncodingConfigWithDefault<E: Default>`: For sinks that have a default `Encoding`.
+//!
+//! Your sink should define some `Encoding` enum that is used as the `E` parameter.
+//!
+//! You can use either of these for a sink! They both implement `EncodingConfiguration`, which you
+//! will need to import as well.
+//!
+//! # Using a configuration
+//!
+//! To use an `EncodingConfig` involves three steps:
+//!
+//!  1. Choose between `EncodingConfig` and `EncodingConfigWithDefault`.
+//!  2. Call `apply_rules(&mut event)` on this config **on each event** just before it gets sent.
+//!
+//! # Implementation notes
+//!
+//! You may wonder why we have both of these types! **Great question.** `serde` works with the
+//! static `*SinkConfig` types when it deserializes our configuration. This means `serde` needs to
+//! statically be aware if there is a default for some given `E` of the config. Since
+//! We don't require `E: Default` we can't always assume that, so we need to create statically
+//! distinct types! Having `EncodingConfigWithDefault` is a relatively straightforward way to
+//! accomplish this without a bunch of magic.
+//!
+// TODO: To avoid users forgetting to apply the rules, the `E` param should require a trait
+//       `Encoder` that defines some `encode` function which this config then calls internally as
+//       part of it's own (yet to be written) `encode() -> Vec<u8>` function.
+
+mod encoding_config;
+pub use encoding_config::EncodingConfig;
+mod encoding_config_with_default;
+pub use encoding_config_with_default::EncodingConfigWithDefault;
+mod inner;
+mod inner_with_default;
+use inner_with_default::InnerWithDefault;
+
+use crate::{event::Value, Event, Result};
+use serde::{Deserialize, Serialize};
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use string_cache::DefaultAtom as Atom;
+
+/// The behavior of a encoding configuration.
+pub trait EncodingConfiguration<E> {
+    fn codec(&self) -> &E;
+    /// If `Some(_)` this configuration will filter out any field not listed.
+    fn only_fields(&self) -> &Option<Vec<Atom>>;
+    /// If `Some(_)` this configuration will filter out any field listed.
+    fn except_fields(&self) -> &Option<Vec<Atom>>;
+    /// If `Some(_)` this configuration will configure the timestamp output.
+    fn timestamp_format(&self) -> &Option<TimestampFormat>;
+
+    fn set_only_fields(&mut self, fields: Option<Vec<Atom>>) -> Option<Vec<Atom>>;
+    fn set_except_fields(&mut self, fields: Option<Vec<Atom>>) -> Option<Vec<Atom>>;
+    fn set_timestamp_format(&mut self, format: Option<TimestampFormat>) -> Option<TimestampFormat>;
```

Comment:
I see a single caller of the `.set_*` functions, and it doesn't make use of the return value. Would a builder pattern be better, or even an init struct pattern? I don't like the use of setter functions on something that will otherwise be constant during its lifetime.

Reply:
Right, that one was basically added for API consistency. I think a builder is better indeed!